### ISSUE TITLE
Cannot represent instants before the program start

### ISSUE THREAD

**Issue by Rodrigodd** [sebcrozet/instant#45]
In one of my projects, I have the necessity of representing an arbitrary instant in time, which include instants before the program start.

The current implementation relies on creating a `Duration` from `performe.now()`. Because ` performe.now()` return the time since the creation of the page/web worker, and because `Duration` cannot represent negative values, the Wasm version of `Instant` can only represent a small range of past instants.

**Comment by daxpedda** [comment]
Is wasn't able to reproduce this and was also not able to find any information on it, could you elaborate how this can happen?

**Comment by Rodrigodd** [comment]
@daxpedda This line panics with 'overflow when subtracting durations' (if your program has not being running for a 100s):

```
let _ = Instant::now() - Duration::from_secs(100);
```
I added [a test to my PR](https://github.com/sebcrozet/instant/pull/46/commits/56118040612712ecc1c354be24be4d24dd760468) to test this.

**Comment by daxpedda** [comment]
Isn't this as intended? [`std::tiime::Instant`] should behave the same, see [source](https://doc.rust-lang.org/1.68.1/src/std/time.rs.html#420-426).

So the behavior seems correct to me.

**Comment by daxpedda** [comment]
Well yes, but that is not behavior that you should rely on. The starting point of the underlying clock is not defined by Rust.

The reason this currently works on Linux is because `std` uses [`clock_gettime`](https://linux.die.net/man/3/clock_gettime):
> All implementations support the system-wide realtime clock, which is identified by CLOCK_REALTIME. Its time represents seconds and nanoseconds since the Epoch.

So the starting time is since the UNIX epoch. So unless your computer starts with a time set to 1970 it will always be much bigger then 0.

On Windows for example, `std` uses [`QueryPerformanceCounter`](https://learn.microsoft.com/en-us/windows/win32/api/profileapi/nf-profileapi-queryperformancecounter), which doesn't start from the UNIX epoch, 0 can be the start of when you boot. So the example you posted above might fail on Windows.

---

That said, looking at your PR, it seems to me that probably the better way to do this is to add `Performance.timeOrigin` to the calculation instead of shifting the whole number to an arbitrary point.

In any case, from a general-purpose library standpoint, an additional calculation is additional overhead for a non-conforming use-case, imho.

**Comment by Rodrigodd** [comment]
Yes, you are right, on Windows I can get my example to panic with a value as little as `20_000` seconds, which matches with the time since boot behavior. So I should not really be relying on that.

I could argue that they should be able to handle negative values (because both Windows and POSIX appear to return signed values), but some platforms may not, and that is also beyond this issue.

So fell free to close this issue.
